### PR TITLE
Standalone dark mode

### PR DESCRIPTION
**Pull request by ivonani**
Two new themes (FlatDarculaLaf and FlatLightLaf) were added.

### CODE REVIEW

**Comment by afdia** [comment]
Hi, most of the code looks fine. I have added some comments as you can see above.

3 more things:

1. about the colors:
Please try to enable "show grid" and "developer mode" and check if the colors look good. As you can see on the following screenshot the grid is too bright to see anything else. It should be barely visible as seen in the light theme
Dark Screenshot: grafik
Light Grid Screenshot: grafik

Also the developer options which show the elements name on the bottom right (only when the element is selected) still use black which makes it unreadable. I guess we have a way to get the "foreground color" maybe use this instead?

Last but not least the yellow marker for the currently selected line in the properties is a bit too bright. If it's easy maybe make the color easier on the eyes when using dark mode

2. there are issues when switching themes. I have Umlet as a relatively small window in the middle of the screen. When I switch it suddently maximizes and the properties panel also maximizes so now the palette is no longer visible. Do you have the same issues?

3. The whole startuphelp.html seems to be too complicated and it didn't work all the time when switching themes (sometimes the HTML content just didn't appeared). Also with this approach, we have to make sure to keep both variants of the files in synch if we want to change the text in the future. As they only differ in the background color, a much better approach would be to make the background transparent or automatically derived from the theme (just as the foreground which is black on light-themes but gray on the dark theme)
a. remove `background-color: white;` from the startuphelp.html and in StartupHelpText.java remove the part which sets the background color of the JComponent (line 102-106)
b. now the background is no longer explicitly specified. It seems that it's derived from the theme just as the foreground color

After this change, we only have one html file and switching themes always shows the html (and the StandaloneGUI.showStartUpText() method would no longer be required)
The only problem ist that while the background color is white on the old umlet themes (which is correct), it is gray on both flatleaf themes. I could live with the gray background (maybe Martin has an issue with it?) but I have the feeling that it should be possible to change the default background color of JComponents for both FlatLeaf themes (the light would get white and the dark would get #282828). Then it should automatically use this derived color without reloading the html panel.

An alternative solution would be:
keep just one startuphelp.html but replace `background-color: white;` with `background-color: #282828;` after loading (at least we wouldn't have to keep 2 files in sync, but the missing-html-on-theme-switch problem would still exist but switching style is not that usual so it would be fine

I would say try to find out of you can change the default background color of such html panels, then we wouldn't need the reload. If it's not easy we could stick with the reload, but please reduce it to one file with an optional background-color replacement after reading the file if necessary

Other than that it looks fine to me, you can fix those things and update the pull request

**Comment by ivonani** [comment]
1. Regarding colours: I fixed grid color, elements name color and color of marker for selected line. For grid and marker I chose colours which I found suitable but I can change to darker shade for grid or to any other color for marker if you think it would look better.
2. Regarding switching themes: I have the same issue, but this issue happens for me even in the master branch when changing themes. Could you check if you also have this issue without my changes?
3. Regarding startuphelp.html: I was able in the end to change background of html panel dynamically, without having the second html file in place, so this should also be fixed :)

**Comment by afdia** [comment]
Looks, good but I would change the dark grid to
`GRID_COLOR_DARK = new Color(70, 70, 70);`

I've tried some values but a darker grid looks odd and if it's too invisible, it's weird to look at it (like an optical illusion) so I guess 70,70,70 would be a good compromise. What do you think?

Also please check the first few comments in this pull request where I commented some lines of code (like setBackground).
If those are fixed and you say that you're done, I will look through the changes once more and rebase+merge them

**Comment by afdia** [comment]
Sorry, that was my mistake (I haven't used Githubs review functionality very much before :) )

Now the review is marked as finished and you should see the comments (before I was seeing them but another user wasn't)

File: umlet-standalone/src/main/java/com/baselet/standalone/gui/StandaloneGUI.java [hunk 1]
```diff
@@ -282,22 +281,74 @@ public void requestFocus() {
...
+			Color background = isDarkMode ? new Color(40, 40, 40) : Color.WHITE;
+			Color foreground = isDarkMode ? Color.LIGHT_GRAY : Color.BLACK;
+
+			if (diagramHandler != null && diagramHandler.getDrawPanel() != null) {
+                showStartUpText(diagramHandler);
+				updatePropertyPaneTheme(propertyPane, background, foreground);
+			}
+
+			if (customElementHandler != null && customElementHandler.getCodePane() != null) {
+				if (customElementHandler.getCodePane() != null) {
```
[the remark this answers is not in the document]
Comment (ivonani) [reply umlet-standalone/src/main/java/com/baselet/standalone/gui/StandaloneGUI.java:303]: You are right I forgot to delete it, but the first check is redundant. It is deleted now

File: umlet-swing/src/main/java/com/baselet/diagram/DrawPanel.java [hunk 2]
```diff
@@ -104,21 +112,38 @@ public void run() {
 		this.repaint(); // repaint the drawpanel to be sure everything is visible (startuphelp etc)
 	}
 
+	@Override
+	public void setBackground(Color bg) {
+		super.setBackground(bg);
+	}
+
```
[the remark this answers is not in the document]
Comment (ivonani) [reply umlet-swing/src/main/java/com/baselet/diagram/DrawPanel.java]: You are right, deleted

File: umlet-swing/src/main/java/com/baselet/diagram/DrawPanel.java [hunk 3]
```diff
@@ -556,6 +581,11 @@ public void updateElements() {
 		}
 	}
 
+	@Override
+	public Color getBackground() {
+		return super.getBackground();
+	}
+
```
[the remark this answers is not in the document]
Comment (ivonani) [reply umlet-swing/src/main/java/com/baselet/diagram/DrawPanel.java]: removed

File: umlet-standalone/src/test/java/com/baselet/standalone/MainBatchmodeTest.java [hunk 4]
```diff
@@ -60,15 +62,21 @@ public void batchConvertToPng_wildcardAndNoOutputParam_newCustomElement() throws
 		File copy = copyInputToTmp("in_newCustomElement.uxf");
 		String wildcard = copy.getParent().toString() + "/*";
 		MainStandalone.main(new String[] { "-action=convert", "-format=png", "-filename=" + wildcard });
-		assertImageEqual(new File(copy + "." + "png"), new File(TEST_FILE_LOCATION + "out_newCustomElement.png"), new File(TEST_FILE_LOCATION + "out_newCustomElement2.png"), new File(TEST_FILE_LOCATION + "out_newCustomElement3.png"));
+		if (!Config.getInstance().getUiManager().equals(Constants.FLAT_DARCULA_THEME)) {
+			assertImageEqual(new File(copy + "." + "png"), new File(TEST_FILE_LOCATION + "out_newCustomElement.png"),
+				new File(TEST_FILE_LOCATION + "out_newCustomElement2.png"),
+				new File(TEST_FILE_LOCATION + "out_newCustomElement3.png"));
+		}
```
[the remark this answers is not in the document]
Comment (ivonani) [reply umlet-standalone/src/test/java/com/baselet/standalone/MainBatchmodeTest.java]: I created config file but it needed to be at location umlet/umlet-standalone/src/test/resources/JUnit/Library/Preferences/UMLet/umlet.cfg in order to be successful. Is that also fine?

**Comment by ivonani** [comment]
I set grid color to GRID_COLOR_DARK = new Color(70, 70, 70); and it looks better now.
I also fixed all commented code lines, so the pull request is ready for review again :)

**Comment by afdia** [comment]
Thanks, I have squashed your commits to 1 and changed 2 things:
1. organized imports to match the eclipse formatter style of umlet
2. the tests with the custom configs were Macos specific so they didn't work for me. I added a new optional explicit sysprop to set the cfg path which is used now

Please check out the latest state of this repo and maybe test once more if everything works, but I would say the dark mode is finished with this thanks :)

**Comment by afdia** [comment]
Unfortunately there is an issue with `ThemeFactory.getCurrentTheme()`

Before your change it was only triggered once (when theme == null was true) and now it triggers on every call which breaks the web-version (it has an infinite loop) and getCurrentTheme intuitively should only initialize the theme once and then keep it (it's a getter and not a setter after all)

You can see the endless loop in the web version by calling the Umletino GWT SuperDev Mode.launch and copying the url from the GWT popup in the browser
Its basically 

```
    at BaseletGWT_1_g$.onThemeChange_0_g$ [as onThemeChange_4_g$] (BaseletGWT.java:110:1)
    at changeTheme_0_g$ (ThemeFactory.java:51:1)
    at getCurrentTheme_0_g$ (ThemeFactory.java:59:1)
    at BaseletGWT_1_g$.onThemeChange_0_g$ [as onThemeChange_4_g$] (BaseletGWT.java:110:1)
...
```

It calls the getCurrentTheme which sets the theme (previously only once, now everytime). By setting the theme, the themechange listeners are triggered which call getCurrentTheme from BaseletGWT, ...

So maybe check if you can switch getCurrentTheme back to init-once-only then the web version will work again

I would suggest to check out the latest version of the umlet repo and start a new pull request from there (so I will leave this one closed), I hope it's just a small change

**Pull request closed without merging**